themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,247 @@ function toggleVisibility(e) {
 	}
 }
 
-function filter(e) {
+function searchOnLoad() {
+	const active = document.getElementsByClassName('active-topic')[0];
+	const search = getSearch();
+	if (active && search) {
+		pageTextSearch(search);
+		tagSearch(search);
+		menuItemsSearch(search);
+	}
+}
+
+function searchOnHashChange() {
+	const search = getSearch();
+	pageTextSearch(search);
+	tagSearch(search);
+	menuItemsSearch(search);
+}
+
+function onSearchChange(e) {
 	e.stopImmediatePropagation();
 	const input = e.target || e.srcElement;
 	const filter = input.value.toLowerCase();
-	const nav = document.getElementById('nav');
-	const li = nav.getElementsByTagName('li');
+	setSearch(filter);
+	updateLinksInMenu();
+}
 
-	for (let i = 0; i < li.length; i++) {
-		const a = li[i].getElementsByTagName('a')[0];
+function setSearch(filter) {
+	if (filter) {
+		location.hash = 'search=' + filter;
+	}
+	else {
+		location.hash = '';
+	}
+}
+
+function getSearch() {
+	const contains = new RegExp(/^#search=/);
+	if (contains.test(location.hash)) {
+		return location.hash.substr(8).replace(/%20/g, ' ').toLowerCase();
+	}
+	return '';
+}
+
+function updateLinksInMenu() {
+	const nav = document.getElementById('nav');
+	const menuItems = nav.getElementsByTagName('li');
+	for (let i = 0; i < menuItems.length; i++) {
+		const a = menuItems[i].querySelector('.title');
+		const dataFound = menuItems[i].getElementsByTagName('data-found')[0];
 		if (a) {
-			if (a.text.toLowerCase().indexOf(filter) > -1) {
-				li[i].style.display = '';
-			} else {
-				li[i].style.display = 'none';
+			let href = a.getAttribute('href');
+			const search = getSearch();
+			if (href.indexOf('#search=') == -1) {
+				href = href + '#search=';
+			}
+			else {
+				href = href.substr(0, href.indexOf('=') + 1);
+			}
+			if (search == '') {
+				href = href.replace(/#search=/, '');
+			}
+			if (dataFound) {
+				const tag = dataFound.getElementsByTagName('a')[0];
+				tag.setAttribute('href', href + search);
+			}
+			a.setAttribute('href', href + search);
+		}
+	}
+}
+
+function pageTextSearch(search) {
+	const h2 = document.getElementsByTagName('h2');
+	const p = document.getElementsByTagName('p');
+	for (let searchTarget of [...h2, ...p]) {
+		const a = searchTarget.getElementsByTagName('a')[0];
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		highlightSymbols(searchTarget, search);
+	}
+	for (let searchTarget of [...h2, ...p]) {
+		const index = searchTarget.textContent.toLowerCase().indexOf(search);
+		if (index > -1 && search != '') {
+			searchTarget.scrollIntoView();
+			return;
+		}
+	}
+}
+
+function tagSearch(search) {
+	const active = document.getElementsByClassName('active-topic')[0];
+	const nav = document.getElementById('nav');
+	const menuItems = nav.getElementsByTagName('li');
+	for (let i = menuItems.length - 1; i >= 0; i--) {
+		const tags = menuItems[i].dataset.tag;
+		if (tags) {
+			const tag = tags.split(',');
+			next: for (let j = 0; j < tag.length; j++) {
```

Comment:
too complicated... goto labels are very limited to use, and this case is not for that